doc/Precision_Guidelines_for_Ngen_Code_Validation.md

```diff
@@ -0,0 +1,36 @@
+
+Usage in NGEN
+
+New model or code implementations will target a code validation at the proposed precision levels for key variables and mass/energy balances (eg, between a standalone model and a modularized model or component). It is not expected that all new formulations will meet the standard (everywhere and in all test timesteps), but checking for high levels of precision will highlight where other factors (some resolvable, some not) may be impacting the portability of code from one simulation context to another.  In other words, precision tolerance usage and checking is not proposed as a show-stopper for development but rather as a diagnostic tool for raising awareness about potential issues to flag and investigate when the project scope permits. 
+
+
+**Experience to Date:  Comparison of Results of NGEN CFE codes with BMI and without BMI**
+
+In our tests of the Conceptual Functional Equivalent (CFE) versus CFE implemented via the Basic Model Interface (BMI) modularization methods, earlier comparisons checked six digits after the decimal points from the formatted outputs for catchment id-87. For these comparisons, we have seen identical values between the BMI-CFE and the original CFE code for Schaake Output Runoff, GIUH Runoff, Nash Lateral Flow, Ground Water Flow, and Total Discharge, for all 720 time steps, with 1 hour time step size. These tests are performed using varying initial conditions: soil_storage_percentage at 0.22, and 0.667; groundwater_storage_percentage at 0.5 and 0.01. We also looked at the total discharge data from standard screen output, which indicated 8-9 digits agreement between BMI-CFE and CFE calculations. In earlier tests, we have also shown similar agreement between CFE calculation results and the original T-Shirt C code calculation results.
+
+
+**Possible reasons for precision discrepancies between CFE and BMI-CFE:**
+
+The double precision numbers have an accuracy of 15 decimal digits, the loss of accuracy from rounding off error is expected, but precisely how many digit loss probably varies and depends on the specific code. Since BMI-CFE code has additional interface and difference in coding details, we hypothesize the memory allocation, cache usage, data sets loading into the registers may be different. These likely affect the rounding off processes and cause the least significant digits to differ. According to some literature, computer hardware architecture, compiler, even optimization may all affect numerical precision. 

```

Comment:
>memory allocation, cache usage, datasets loading into the registers

These should have no effect on rounding error,  those are caused by operations on on the floats.  Copying bits, in the absence of bit flip or bit rot, should always give the same the bits, hence the same precision.  The only time this would happen under normal circumstances is casting between single/double precision.  If the interfaces don't do this, they shouldn't the be cause of precision loss.  But simply doing operations in different orders can definitely cause differences in rounding error and hence precsion.

Reply:
My thinking was, taking for example array operation, if in one case you load elements 1 to 99, do some calculations and save the result into memory. In another case, you load element 2 to 100, and do some calculations, save the result into memory, should the round off error be different?

Could you think of some good reasons these might have a role in round off error? If not we need to delete it.

Reply:
@donaldwj toughts?